api/src/test/java/org/openmrs/module/idgen/LocationBasedPrefixProviderTest.java

```diff
@@ -0,0 +1,138 @@
+		LocationAttribute location5PrefixAtt = location5.getActiveAttributes().iterator().next();
+		// Invalidate prefix attributes for locations found in the middle of the tree
+		locationA2PrefixAtt.setValue(" ");
+		location5PrefixAtt.setValue(" ");
+		Assert.assertEquals("LOC-3", locationPrefixProvider.getLocationPrefix(locationA3));
+	}
+	
+	@Test(expected = RuntimeException.class)
+	public void getLocationPrefix_throwAnExceptionIfNoValidPrefixIsFound() {
+		// Invalidate valid prefixes
+		LocationAttribute location2PrefixAtt = location3.getActiveAttributes().iterator().next();
+		location2PrefixAtt.setValue(" ");
+		locationPrefixProvider.getLocationPrefix(location3);
+	}
+	
+	@Test
+	public void getLocationPrefixRecursively_shouldPickThePrefixFromCurrentLocationIfOneIsSet() {
+		Assert.assertEquals("LOC-5", locationPrefixProvider.getLocationPrefix(location5));
+	}
+	
+	private void setupLocationTree() {
+		Location location1 = new Location(); 
+		Location location2 = new Location();
+		location2.setParentLocation(location1);
+		
+		location3 = new Location(); 
+		location3.setParentLocation(location2);
+		LocationAttribute location3PrefixAtt = new LocationAttribute();
+		location3PrefixAtt.setAttributeType(createPrefixAttributeType());
+		location3PrefixAtt.setValue("LOC-3");
+		location3.addAttribute(location3PrefixAtt);
+		
+		Location location4 = new Location();
+		location4.setParentLocation(location3);
+		
+		location5 = new Location(); 
+		location5.setParentLocation(location4);
+		LocationAttribute location5PrefixAtt = new LocationAttribute();
+		location5PrefixAtt.setAttributeType(createPrefixAttributeType());
+		location5PrefixAtt.setValue("LOC-5");
+		location5.addAttribute(location5PrefixAtt);
+		
+		// First branch
+		Location locationA1 = new Location();
+		location5.addChildLocation(locationA1);
+		
+		locationA2 = new Location(); 
+		locationA2.setParentLocation(locationA1);
+		LocationAttribute locationA2PrefixAtt = new LocationAttribute();
+		locationA2PrefixAtt.setAttributeType(createPrefixAttributeType());
+		locationA2PrefixAtt.setValue("LOC-A2");
+		locationA2.addAttribute(locationA2PrefixAtt);
+		
+		locationA3 = new Location();
+		locationA3.setParentLocation(locationA2);
+		
+		// Second Branch
+		Location locationB1 = new Location();
+		location5.addChildLocation(locationB1);
+
+		Location locationB2 = new Location();
+		locationB2.setParentLocation(locationB1);
+		
+		locationB3 = new Location();
+		locationB3.setParentLocation(locationB2);
+
+	}
+		
+	private LocationAttributeType createPrefixAttributeType() {
+		LocationAttributeType prefixAttrType = new LocationAttributeType();
+		prefixAttrType.setName(IdgenConstants.PREFIX_LOCATION_ATTRIBUTE_TYPE);
+		prefixAttrType.setMinOccurs(0);
+		prefixAttrType.setMaxOccurs(5);
+		prefixAttrType.setDatatypeClassname("org.openmrs.customdatatype.datatype.FreeTextDatatype");
+		return prefixAttrType;
```

Comment:
You should reuse the activator method that creates the metadata for the module.
Also this doesn't make sense:
```java
prefixAttrType.setMaxOccurs(5);
```

Reply:
We discussed this before @mks-d and we agreed on ignoring this. We can't reuse the activator method cause it a DB transaction method. This test class is `@RunWith(PowerMockRunner.class)` so I'm not sure why we can't `extend BaseModuleContextSensitiveTest` with tests run by PowerMockRunner. Can't we leave this out?